# COMS 4995 HW4
#### Jianfeng Zhuang: JZ3872
#### Huazhang Liu: HL3338

### Task 1

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math
from scipy.stats import randint

In [12]:
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder, MinMaxScaler, PolynomialFeatures
from sklearn.compose import ColumnTransformer
from sklearn.compose import make_column_transformer
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.model_selection import cross_val_score, GridSearchCV, KFold
from sklearn.impute import SimpleImputer
from category_encoders import TargetEncoder
from category_encoders.leave_one_out import LeaveOneOutEncoder
from category_encoders.woe import WOEEncoder
from sklearn.impute import KNNImputer
from sklearn.ensemble import RandomForestRegressor
from xgboost.sklearn import XGBRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import cross_val_score

In [3]:
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.linear_model import ElasticNet

In [10]:
wine = pd.read_csv("winemag-data-130k-v2.csv")
wine = wine.drop("Unnamed: 0", axis = 1)
wine.head()

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

In [11]:
wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 13 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   country                129908 non-null  object 
 1   description            129971 non-null  object 
 2   designation            92506 non-null   object 
 3   points                 129971 non-null  int64  
 4   price                  120975 non-null  float64
 5   province               129908 non-null  object 
 6   region_1               108724 non-null  object 
 7   region_2               50511 non-null   object 
 8   taster_name            103727 non-null  object 
 9   taster_twitter_handle  98758 non-null   object 
 10  title                  129971 non-null  object 
 11  variety                129970 non-null  object 
 12  winery                 129971 non-null  object 
dtypes: float64(1), int64(1), object(11)
memory usage: 12.9+ MB


In [37]:
wine_US = wine[wine['country'] == "US"]
wine_US = wine_US.drop('country', axis = 1)
wine_US.index = np.arange(0, wine_US.shape[0])
wine_US

description  \
0      Tart and snappy, the flavors of lime flesh and...   
1      Pineapple rind, lemon pith and orange blossom ...   
2      Much like the regular bottling from 2012, this...   
3      Soft, supple plum envelopes an oaky structure ...   
4      Slightly reduced, this wine offers a chalky, t...   
...                                                  ...   
54499  Hailing from one of the more popular vineyards...   
54500  There's no bones about the use of oak in this ...   
54501  This opens with herbaceous dollops of thyme an...   
54502  This Zinfandel from the eastern section of Nap...   
54503  Citation is given as much as a decade of bottl...   

                              designation  points  price    province  \
0                                     NaN      87   14.0      Oregon   
1                    Reserve Late Harvest      87   13.0    Michigan   
2      Vintner's Reserve Wild Child Block      87   65.0      Oregon   
3                          Mountain Cuvée      87   19.0  California   
4                                     NaN      87   34.0  California   
...                                   ...     ...    ...         ...   
54499    Jurassic Park Vineyard Old Vines      90   20.0  California   
54500                    Barrel Fermented      90   35.0  California   
54501                       Blocks 7 & 22      90   35.0  California   
54502                                 NaN      90   22.0  California   
54503                                 NaN      90   75.0      Oregon   

                  region_1           region_2         taster_name  \
0        Willamette Valley  Willamette Valley        Paul Gregutt   
1      Lake Michigan Shore                NaN  Alexander Peartree   
2        Willamette Valley  Willamette Valley        Paul Gregutt   
3              Napa Valley               Napa      Virginie Boone   
4         Alexander Valley             Sonoma      Virginie Boone   
...                    ...                ...                 ...   
54499    Santa Ynez Valley      Central Coast       Matt Kettmann   
54500          Napa Valley               Napa      Virginie Boone   
54501          Napa Valley               Napa      Virginie Boone   
54502        Chiles Valley               Napa      Virginie Boone   
54503               Oregon       Oregon Other        Paul Gregutt   

      taster_twitter_handle  \
0               @paulgwine    
1                       NaN   
2               @paulgwine    
3                   @vboone   
4                   @vboone   
...                     ...   
54499         @mattkettmann   
54500               @vboone   
54501               @vboone   
54502               @vboone   
54503           @paulgwine    

                                                   title             variety  \
0          Rainstorm 2013 Pinot Gris (Willamette Valley)          Pinot Gris   
1      St. Julian 2013 Reserve Late Harvest Riesling ...            Riesling   
2      Sweet Cheeks 2012 Vintner's Reserve Wild Child...          Pinot Noir   
3      Kirkland Signature 2011 Mountain Cuvée Caberne...  Cabernet Sauvignon   
4      Louis M. Martini 2012 Cabernet Sauvignon (Alex...  Cabernet Sauvignon   
...                                                  ...                 ...   
54499  Birichino 2013 Jurassic Park Vineyard Old Vine...        Chenin Blanc   
54500  Flora Springs 2013 Barrel Fermented Chardonnay...          Chardonnay   
54501  Hendry 2012 Blocks 7 & 22 Zinfandel (Napa Valley)           Zinfandel   
54502             Houdini 2011 Zinfandel (Chiles Valley)           Zinfandel   
54503                  Citation 2004 Pinot Noir (Oregon)          Pinot Noir   

                   winery  
0               Rainstorm  
1              St. Julian  
2            Sweet Cheeks  
3      Kirkland Signature  
4        Louis M. Martini  
...                   ...  
54499           Birichino  
54500       Flora Springs  
54501              Hendry  
54502           

From the above, we know that the target column is "points'. 

In [38]:
wine_US['designation'].unique().shape

(14184,)

In [39]:
wine_US[wine_US['designation'] == "Reserve"]

description designation  points  \
91     There is an odd, piercing edge to the aromas, ...     Reserve      85   
98     Tasty, with pie-filling cherries, blackberries...     Reserve      85   
125    With a fruity-sugary finish, this Zin will app...     Reserve      86   
161    Fine acidity and a bracing minerality, like a ...     Reserve      92   
226    This is a lovely wine, classically rendered in...     Reserve      93   
...                                                  ...         ...     ...   
54412  This robust wine is made from 62% Merlot, 37% ...     Reserve      90   
54420  Blackberry juice mixes with graphite, cedar an...     Reserve      90   
54462  Dusty tannins make for a soft texture in this ...     Reserve      90   
54475  This wine is mostly Cabernet Sauvignon (78%), ...     Reserve      90   
54491  This ripe, rich, almost decadently thick wine ...     Reserve      91   

       price    province                   region_1           region_2  \
91      28.0      Oregon          Willamette Valley  Willamette Valley   
98      25.0  California            Sonoma Mountain             Sonoma   
125     34.0  California                  Mendocino                NaN   
161     40.0  California                   Carneros        Napa-Sonoma   
226     70.0  California            Howell Mountain               Napa   
...      ...         ...                        ...                ...   
54412  140.0  California  Sonoma County-Napa County        Napa-Sonoma   
54420   50.0  California          Santa Ynez Valley      Central Coast   
54462   44.0  California                   Carneros        Napa-Sonoma   
54475   60.0  Washington       Columbia Valley (WA)    Columbia Valley   
54491  105.0  Washington    Walla Walla Valley (WA)    Columbia Valley   

            taster_name taster_twitter_handle  \
91         Paul Gregutt           @paulgwine    
98                  NaN                   NaN   
125                 NaN                   NaN   
161                 NaN                   NaN   
226      Virginie Boone               @vboone   
...                 ...                   ...   
54412    Virginie Boone               @vboone   
54420     Matt Kettmann         @mattkettmann   
54462    Virginie Boone               @vboone   
54475  Sean P. Sullivan         @wawinereport   
54491      Paul Gregutt           @paulgwine    

                                                   title  \
91     Silvan Ridge 2006 Reserve Pinot Noir (Willamet...   
98            Work 2004 Reserve Merlot (Sonoma Mountain)   
125          Brutocao 2006 Reserve Zinfandel (Mendocino)   
161    Robert Mondavi 2011 Reserve Chardonnay (Carneros)   
226    W.H. Smith 2012 Reserve Cabernet Sauvignon (Ho...   
...                                                  ...   
54412  Pride Mountain 2013 Reserve Claret (Sonoma Cou...   
54420  Brander 2014 Reserve Cabernet Sauvignon (Santa...   
54462            Ca' Momi 2013 Reserve Merlot (Carneros)   
54475   Matthews 2012 Reserve Red (Columbia Valley (WA))   
54491  Nicholas Cole Cellars 2004 Reserve Red (Walla ...   

                        variety                 winery  
91                   Pinot Noir           Silvan Ridge  
98                       Merlot                   Work  
125                   Zinfandel               Brutocao  
161                  Chardonnay         Robert Mondavi  
226          Cabernet Sauvignon             W.H. Smith  
...                         ...                    ...  
54412                    Claret         Pride Mountain  
54420        Cabernet Sauvignon                Brander  
54462                    Merlot               Ca' Momi  
54475  Bordeaux-style Red Blend               Matthews  
54491                 Red Blend  Nicholas Cole Cellars  

[1485 rows x 12 columns]

In [58]:
wine_US['region_1'].unique().shape

(265,)

However, since the information contained in this column is complicated, including the city names, county names, even the names of the states, it should be used as text. 

In [56]:
wine_US['region_2'].unique().shape

(18,)

In [61]:
sum(pd.isna(wine_US['region_2']))/len(wine_US)

0.07326067811536768

In [63]:
sum(pd.isna(wine_US['region_2']))

3993

In [53]:
wine_US[wine_US['region_2'] == "Napa"]

description  \
3      Soft, supple plum envelopes an oaky structure ...   
20     This shows jelly-like flavors of orange and pe...   
23     Syrupy and dense, this wine is jammy in plum a...   
30     Juicy plum, raspberry and pencil lead lead the...   
31     This is the producer's first release of this w...   
...                                                  ...   
54483  Aged entirely in stainless steel, this combine...   
54485  Pencil shaving and graphite notes lend a class...   
54500  There's no bones about the use of oak in this ...   
54501  This opens with herbaceous dollops of thyme an...   
54502  This Zinfandel from the eastern section of Nap...   

                designation  points  price    province         region_1  \
3            Mountain Cuvée      87   19.0  California      Napa Valley   
20           Estate Bottled      85   30.0  California      Napa Valley   
23                   Estate      86  100.0  California      Napa Valley   
30     Bella Vetta Vineyard      86   75.0  California  Howell Mountain   
31             Estate Grown      86   55.0  California        Calistoga   
...                     ...     ...    ...         ...              ...   
54483           Proprietary      91   30.0  California      Napa Valley   
54485                   NaN      91   35.0  California      Napa Valley   
54500      Barrel Fermented      90   35.0  California      Napa Valley   
54501         Blocks 7 & 22      90   35.0  California      Napa Valley   
54502                   NaN      90   22.0  California    Chiles Valley   

      region_2     taster_name taster_twitter_handle  \
3         Napa  Virginie Boone               @vboone   
20        Napa             NaN                   NaN   
23        Napa  Virginie Boone               @vboone   
30        Napa  Virginie Boone               @vboone   
31        Napa  Virginie Boone               @vboone   
...        ...             ...                   ...   
54483     Napa  Virginie Boone               @vboone   
54485     Napa  Virginie Boone               @vboone   
54500     Napa  Virginie Boone               @vboone   
54501     Napa  Virginie Boone               @vboone   
54502     Napa  Virginie Boone               @vboone   

                                                   title             variety  \
3      Kirkland Signature 2011 Mountain Cuvée Caberne...  Cabernet Sauvignon   
20     RustRidge 2010 Estate Bottled Chardonnay (Napa...          Chardonnay   
23     Okapi 2013 Estate Cabernet Sauvignon (Napa Val...  Cabernet Sauvignon   
30     Hindsight 2013 Bella Vetta Vineyard Cabernet S...  Cabernet Sauvignon   
31     Hindsight 2012 Estate Grown Petite Sirah (Cali...        Petite Sirah   
...                                                  ...                 ...   
54483      Paraduxx 2015 Proprietary White (Napa Valley)         White Blend   
54485   Conn Creek 2013 Cabernet Sauvignon (Napa Valley)  Cabernet Sauvignon   
54500  Flora Springs 2013 Barrel Fermented Chardonnay...          Chardonnay   
54501  Hendry 2012 Blocks 7 & 22 Zinfandel (Napa Valley)           Zinfandel   
54502             Houdini 2011 Zinfandel (Chiles Valley)           Zinfandel   

                   winery  
3      Kirkland Signature  
20              RustRidge  
23                  Okapi  
30              Hindsight  
31              Hindsight  
...                   ...  
54483            Paraduxx  
54485          Conn Creek  
54500       Flora Springs  
54501              Hendry  
54502             Houdini  

[6814 rows x 12 columns]

Since there are are total of 14184 unique values of designations, and there are 1485 observations with designation "Reserve", this column is considered as a text column.  

In [42]:
wine_US['taster_name'].unique().shape

(16,)

In [50]:
wine_US['title'].unique().shape

(50229,)

In [49]:
wine_US['variety'].unique().shape

(257,)

In [45]:
wine_US['winery'].unique().shape

(5375,)

Since the column "Winery" consists of 5375 rows, and mostly composited of names, it should be used as a categorical variable through target encoding. 

In [64]:
wine_US_1 = wine_US.drop(['description', 'designation', 'region_1', "taster_twitter_handle", "title"], axis = 1)

In [65]:
wine_US_1.head()

points  price    province           region_2         taster_name  \
0      87   14.0      Oregon  Willamette Valley        Paul Gregutt   
1      87   13.0    Michigan                NaN  Alexander Peartree   
2      87   65.0      Oregon  Willamette Valley        Paul Gregutt   
3      87   19.0  California               Napa      Virginie Boone   
4      87   34.0  California             Sonoma      Virginie Boone   

              variety              winery  
0          Pinot Gris           Rainstorm  
1            Riesling          St. Julian  
2          Pinot Noir        Sweet Cheeks  
3  Cabernet Sauvignon  Kirkland Signature  
4  Cabernet Sauvignon    Louis M. Martini